confluent_kafka/schema_registry/serdes.py

```diff
@@ -0,0 +1,264 @@
+from confluent_kafka.serialization.kafka_serdes import Serializer, Deserializer
+
+_MAGIC_BYTE = 0
+
+
+class SubjectNameStrategy(object):
+    """
+    Instructs the Serializer on how to construct subject names when registering
+    schemas. See the Schema Registryy documentation for additional details.
+
+    .. _SchemaRegistry Documentation:
+        https://docs.confluent.io/current/schema-registry/serializer-formatter.html#group-by-topic-or-other-relationships
+
+    """
+
+    # Configuration enum-like structure
+    TopicNameStrategy = 0
+    TopicRecordNameStrategy = 1
+    RecordNameStrategy = 2
+
+    _str = ['TopicNameStrategy', 'TopicRecordNameStrategy', 'RecordNameStrategy']
+
+    def __call__(self, schema, ctx):
+        raise NotImplementedError
+
+    @classmethod
+    def __str__(cls, strategy):
+        return cls._str[strategy]
+
+
+class TopicNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {topic}={key|value}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+    def __call__(self, schema, ctx):
+        return ctx.topic + "-" + str(ctx.field)
+
+
+class TopicRecordNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {topic}={record.name}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+    def __call__(self, schema, ctx):
+        return ctx.topic + "-" + schema.name
+
+
+class RecordNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {record.name}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+    def __call__(self, schema, ctx):
+        return schema.name
+
+
+class ContextStringIO(io.BytesIO):
+    """
+    Wrapper to allow use of StringIO via 'with' constructs.
+
+    """
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+        return False
+
+
+class AvroSerializer(Serializer):
+    """
+    AvroSerializer encodes objects into the Avro binary format in accordance
+    with the schema provided at instantiation.
+
+    Unless configured otherwise the AvroSerializer will automatically register
+    its schema with the provided registry instance.
+
+    Arguments:
+        conf (AvroSerializerConfig): Avro Serializer configuration instance
+        schema (Schema): Parsed Avro Schema object
+
+    .. _Schema Registry specification
+        https://avro.apache.org/docs/current/spec.html for more details.
+
+    .. _Schema Resolution
+        https://avro.apache.org/docs/1.8.2/spec.html#Schema+Resolution
+
+    """
+    def __init__(self, conf, schema):
+        self.registry = SchemaRegistryClient(conf.SchemaRegistryConfig)
```

Comment:
In .NET I pass this in as a constructor argument, though in Python i think this is less critical.

The main reason it makes sense in .net is the sr client needs to be explicitly disposed, so if internal to the serializer, the serializer would need to be disposable. then you need to worry about documenting / user worrying about ownership of the serializer (does it pass to the producer or can it be reused?). Anyway, in python, i believe it's perfectly fine to rely on the destructor for this.

Another reason is it's slightly more efficient to just have one SRC instance.

Another reason is i think it's nice to separate the concept of SR and Serializer. Firstly, in the config, secondly i think although more verbose, having the SR instance is descriptive to see in program flow.

Anyway, probably not a big deal to leave as is, but I err on the side of passing an SR instance in.